src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -0,0 +1,230 @@
+    : segment_provider(sp), segment_manager(sm) {}
+  write_iertr::future<> write(std::list<LogicalCachedExtentRef>& extent) final;
+
+private:
+  bool _needs_roll(segment_off_t length) const;
+  roll_segment_ertr::future<> roll_segment();
+  init_segment_ertr::future<> init_segment(Segment& segment);
+
+  using extents_to_write_t = std::vector<LogicalCachedExtentRef>;
+  void add_extent_to_write(
+    ool_record_t&,
+    LogicalCachedExtentRef& extent);
+
+  SegmentProvider& segment_provider;
+  SegmentManager& segment_manager;
+  SegmentRef current_segment;
+  std::vector<SegmentRef> open_segments;
+  segment_off_t allocated_to = 0;
+};
+
+using SegmentedOolWriterRef = std::unique_ptr<SegmentedOolWriter>;
+
+class scan_valid_records_cursor;
+
+template <typename HintT = empty_hint_t>
+class ExtentAllocator {
+public:
+  using scan_device_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  virtual CachedExtentRef alloc_ool_extent(
+    Transaction& t,
+    extent_types_t type,
+    const HintT hint,
+    segment_off_t length) = 0;
+
+  virtual ~ExtentAllocator() {};
+};
+
+template <typename HintT = empty_hint_t>
+using ExtentAllocatorRef = std::unique_ptr<ExtentAllocator<HintT>>;
+
+template <typename IndexT, typename HintT = empty_hint_t>
+class SegmentedAllocator : public ExtentAllocator<HintT> {
+public:
+  using calc_target_func_t = typename std::function<IndexT (HintT)>;
+  using alloc_extent_ertr = SegmentedOolWriter::alloc_extent_ertr;
+  SegmentedAllocator(
+    SegmentProvider& sp,
+    SegmentManager& sm,
+    Cache& cache,
+    calc_target_func_t&& calc_target_func)
+    : segment_provider(sp),
+      segment_manager(sm),
+      cache(cache),
+      calc_target_func(std::move(calc_target_func))
+  {}
+
+  CachedExtentRef alloc_ool_extent(
+    Transaction& t,
+    extent_types_t type,
+    const HintT hint,
+    segment_off_t length) final {
+    auto index = calc_target_func(hint);
+    auto iter = writers.find(index);
+    if (iter == writers.end()) {
+      iter = writers.emplace(
+        index,
+        std::make_unique<SegmentedOolWriter>(
+          segment_provider,
+          segment_manager)).first;
+    }
+    auto& writer = iter->second;
+
+    auto nextent = cache.alloc_new_extent_by_type(
+      t, type, length, paddr_t{ZERO_SEG_ID, fake_paddr_off});
+    fake_paddr_off += length;
+    nextent->extent_writer = writer.get();
+    return nextent;
+
+  }
+
+  using read_segment_header_ertr = crimson::errorator<
+    crimson::ct_error::enoent,
+    crimson::ct_error::enodata,
+    crimson::ct_error::input_output_error
+    >;
+  using read_segment_header_ret = read_segment_header_ertr::future<
+    segment_header_t>;
+  read_segment_header_ret read_segment_header(segment_id_t segment);
+
+private:
+  segment_off_t fake_paddr_off = 0;
+
+  SegmentProvider& segment_provider;
+  SegmentManager& segment_manager;
+  std::map<IndexT, SegmentedOolWriterRef> writers;
```

Comment:
> I think we only need one simple map `EPM -[hint]-> ExtentAllocater` directly.

`EPM -[index]-> Device(SegmentedAllocator)` is actually `EPM -[index]-> Device(SegmentedAllocator)`. SegmentedOolWriter is only used by SegmentedOolAllocator, RBM and PMem don't need to worry about it.

Reply:
I meant we can flatten the second `SegmentedAllocator -[another-index]-> SegmentedWriter`. So that a hint can be directly  indexed to an Allocator. `SegmentedAllocator` may share the same segment_provider/manager, and they can ask for segments from the same segment_provider/manager independently.

Reply:
I don't really want to bother with any of this right now.  I've got a branch below that removes the IndexT and HintT templates and just leaves a ool_placement_hint_t argument to the interfaces as a hint for later.